### PR TITLE
unrealirc_3281_backdoor vulnerability refactored.

### PR DESCRIPTION
**Pull request by thomashaw**
Vulnerable code now held within the service module and conditionally run based on whether the unrealirc_3281_backdoor class has been included. Reduced duplicate code.

Planning on following this pattern for misconfiguration vulnerabilities as modifying the same resources from different classes/namespaces within puppet causes issues. 

The issue I was having was that I could not change the parameter value of the filename parameter within the service/unrealirc module from the vulnerability module. After reading up the general consensus is that puppet modules should be self contained (+ this helps avoid scope issues).

The module now runs unrealirc::vulnerabilities subclass before installation and conditionally applies the changes based on whether the vulnerability (unrealirc_3281_backdoor) has been defined. 

The class definition for unrealirc_3281_backdoor is now just an empty class, the internal changes are kept within the unrealirc::vulnerabilities subclass in the service module.

Let me know what you think.

### CODE REVIEW

**Comment by Jjk422** [comment]
I understand the points made, however I don't think service modules should have any vulnerability dependent code included for numerous reasons.
A couple of reasons for including service/vulnerability only routines in the respective modules are:
1. Adding more space/increasing service module size
2. Service modules should only contain code necessary for that service to start with basic/default settings
3. All changes to configuration should be contained in vulnerability modules separate for ease of access, if we have 10 different software/vulnerabilities for one service installation, having all the vulnerabilities crammed into one service module will be extremely difficult to maintain.

Looking through the code I understand why you have done this now, it may be beneficial to have a utility/system module (stored in the same module folder as a user creation module and a breadcrumb insertion module) that extracts and sorts files that will pass the extracted data back to the module, although we need parametrisation for this really.

However, puppet modules are supposed to be self contained, this is true but mostly due to things like portability and ease of use, e.g. If I want a puppet module I don't want to download 10 other ones to make it work or have to know the internal structure of the module.
This shouldn't really affect SecGen though as the installation should be handled in one service module to get the basic settings, after that it should be modifiable.
Service/vulnerability config should be performed via vulnerability modules with values from the service modules being passed via facter or the parametrisation methods that are implemented.

A small outline of how I think the machine modules should be in a system:
Get facter tags via tag e.g. <uitlity facter="variable_name::value variable_name2::value2>
\/\/\/\/\/
User_module >> Service_module >> Extra_service_config or Vulnerability config
or
User_module >> Vulnerability module >> Extra_service_config or Vulnerability config

This allows for vulnerability config to be easily interchangeable, e.g. I can insert and delete vulnerabilities that change the service without ever touching the service module.
Puppet is not really meant for how we are using it, it is really meant for continuous support for the client systems from a master system, not an install and be done with it.

In summary, I think modules should chain and pass data from one to another (obviously puppet isn't really made for this but parametrisation and facter may make dynamic module variable passing possible)

At least this is my opinion on how modules should work and my understanding of SecGen and puppet.
Just my thoughts on the issue because we want easy to maintain code.
Sorry for the response, it seems a bit disjointed and all over the place,

**Comment by thomashaw** [comment]
Good points, taken on board. The problem was that this vulnerability is from a backdoored version of the Unreal tarball that the service module uses so the changes in the vulnerability module, i.e the selection of the vulnerable tar, need to be handled before installation. 

I agree it makes sense to separate. I've reworked it so there is no vulnerability code in the service module but there is a way of changing the ordering through the require on line 3 of vulnerabilities.pp. the require statement changes the order of execution and the unrealirc_3281_backdoor class's contents are run here instead of later.

The unrealirc_3281_backdoor class' code now just copies the vulnerable .tar.gz in before the unrealirc::install class is run. the file reference in unrealirc::install only copies the secure tarball if the file doesn't already exist.

Let me know what you think, cheers!

**Comment by Jjk422** [comment]
Fully tested and exploitable.
I do like the new vulnerability separation as 3 lines to connect the vulnerability in the service module isn't that bad, plus we can always optimize it later by maybe using the encoding modules in the new parameterization branch when that gets merged.
I'm happy to merge it now if you don't have anything else to do on it.

**Comment by thomashaw** [comment]
nice one, thank you :) nothing else to go into this one so going to merge it now, cheers.
